mtags/src/main/scala-3/scala/meta/internal/pc/ScalaPresentationCompiler.scala

```diff
@@ -490,4 +500,135 @@ case class ScalaPresentationCompiler(
+          short.name,
+          defn.AnyType,
+          EmptyFlags,
+          EmptyFlags,
+          NoSourcePosition
+        )
+        foundTpe
+      }})
+
+    def tryShortenName(short: ShortName)(using Context): Boolean = {
+      history.get(short.name) match {
+        case Some(ShortName(_, other)) => true
+        case None =>
+          val foundTpe = lookupSymbol(short)
+          val syms = List(foundTpe.termSymbol, foundTpe.typeSymbol)
+          val isOk = syms.filter(_ != NoSymbol) match {
+            case Nil => false
+            case founds => founds.exists(_ == short.symbol)
+          }
+          if (isOk) {
+            history(short.name) = short
+            true
+          } else {
+            false
+          }
+      }
+    }
+
+    def tryShortenName(name: Option[ShortName])(using Context): Boolean =
+      name match {
+        case Some(short) => tryShortenName(short)
+        case None => false
+      }
+  }
+
+  /**
+   * Shorten the long (fully qualified) type to shorter representation, so printers
+   * can obtain more readable form of type like `SrcPos` instead of `dotc.util.SrcPos`
+   * (if the name can be resolved from the context).
+   * 
+   * For example,
+   * when the longType is like `TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class dotc)),module util),SrcPos)`,
+   * if `dotc.util.SrcPos` found from the scope, then `TypeRef(NoPrefix, SrcPos)`
+   * if not, and `dotc.util` found from the scope then `TypeRef(TermRef(NoPrefix, module util), SrcPos)`
+   * 
+   * @see Scala 3/Internals/Type System https://dotty.epfl.ch/docs/internals/type-system.html
+   */
+  private def shortType(longType: Type, history: ShortenedNames)(using ctx: Context): Type = {
+    val isVisited = collection.mutable.Set.empty[(Type, Option[ShortName])]
+    val cached = new ju.HashMap[(Type, Option[ShortName]), Type]()
+
+    def loop(tpe: Type, name: Option[ShortName]): Type = {
+      val key = tpe -> name
+      // NOTE: Prevent infinite recursion, see https://github.com/scalameta/metals/issues/749
+      if (isVisited(key)) return cached.getOrDefault(key, tpe)
+      isVisited += key
+      val result = tpe match {
+        case TypeRef(prefix, designator) =>
+          // designator is not necessarily an instance of `Symbol` and it's an instance of `Name`
+          // this can be seen, for example, when we are shortening the signature of 3rd party APIs.
+          val sym =
+            if (designator.isInstanceOf[Symbol]) designator.asInstanceOf[Symbol]
+            else tpe.typeSymbol
+          val short = ShortName(sym)
+          TypeRef(loop(prefix, Some(short)), sym)
+
+        case TermRef(prefix, designator) =>
+          val sym =
+            if (designator.isInstanceOf[Symbol]) designator.asInstanceOf[Symbol]
+            else tpe.termSymbol
+          val short = ShortName(sym)
+          if (sym.isOneOf(Flags.Package | Flags.Module) && history.tryShortenName(name)) NoPrefix
```

Comment:
Why `sym.isOneOf(Flags.Package | Flags.Module) ` is needed? Do you have examples of this in the tests?

Reply:
Aw, realized actually we don't need it!
Besides, we don't need to pass `Some(short)` to `loop` in the else clause (since we already know the `sym` cannot be resolved from the scope in the else clause).

```diff
diff --git a/mtags/src/main/scala-3/scala/meta/internal/pc/ScalaPresentationCompiler.scala b/mtags/src/main/scala-3/scala/meta/internal/pc/ScalaPresentationCompiler.scala
index cfb3e1ca..2298a0ff 100644
--- a/mtags/src/main/scala-3/scala/meta/internal/pc/ScalaPresentationCompiler.scala
+++ b/mtags/src/main/scala-3/scala/meta/internal/pc/ScalaPresentationCompiler.scala
@@ -588,8 +588,8 @@ case class ScalaPresentationCompiler(
             if (designator.isInstanceOf[Symbol]) designator.asInstanceOf[Symbol]
             else tpe.termSymbol
           val short = ShortName(sym)
-          if (sym.isOneOf(Flags.Package | Flags.Module) && history.tryShortenName(name)) NoPrefix
-          else TermRef(loop(prefix, Some(short)), sym)
+          if (history.tryShortenName(name)) NoPrefix
+          else TermRef(loop(prefix, None), sym)
 
         case ThisType(tyref) =>
           if (history.tryShortenName(name)) NoPrefix
```

Reply:
Looks like it's take from the previous implementation, we can leave that in.

Reply:
> Looks like it's take from the previous implementation, we can leave that in.

Actually, yes I took it from the previous implementation and didn't realize we don't need it 🤦
I think it's better to remove it at least now so that we won't be confused with this (like why we need this condition?) in the future.

Reply:
Ok removed that, thanks!